src/engine.ml

```diff
@@ -1596,89 +1637,64 @@ let find_channel state key op =
-                  | Some keys ->
-                      let ch = received_packet ch payload in
-                      let bad_mac hmac =
-                        `Bad_mac (state, hmac, (key, `Data payload))
-                      in
-                      let hmac_algorithm = Config.get Auth state.config in
-                      let+ payload =
-                        incoming_data bad_mac keys hmac_algorithm
-                          state.session.compress payload
-                      in
-                      let act = merge_payload act payload in
-                      (set_ch state ch, out, act))
               | _, Client_static _ ->
-                  (* non-data packet for static client *)
-                  (* XXX(reynir): should we rather fail!? *)
-                  Log.warn (fun m ->
-                      m
-                        "Static client received control channel packet; \
-                         ignoring.");
-                  Ok (state, out, act)
-              | Packet.Hard_reset_client_v3, Client_tls_auth _
-              | Packet.Hard_reset_client_v3, Server_tls_auth _ ->
-                  Error (`No_transition (ch, op, payload))
-              | ( Packet.Ack,
-                  ( Client_tls_auth { tls_auth; _ }
-                  | Server_tls_auth { tls_auth; _ } ) ) -> (
-                  let hmac_len =
-                    Mirage_crypto.Hash.digest_size tls_auth.hmac_algorithm
+                  (* This function is not meant to be called from static client *)
+                  assert false
+              | Packet.Data_v1, _ ->
+                  let+ state, payload =
+                    received_data state key ch set_ch payload
                   in
-                  let* ack = Packet.decode_ack ~hmac_len payload in
-                  let p = `Ack ack in
-                  let bad_mac hmac = `Bad_mac (state, hmac, (key, p)) in
-                  match
-                    let* () =
-                      check_control_integrity bad_mac key p
-                        tls_auth.hmac_algorithm tls_auth.their_hmac
-                    in
-                    expected_packet state.session ch.transport p
-                  with
-                  | Error e ->
-                      (* XXX: only in udp mode? *)
-                      Log.warn (fun m -> m "ignoring bad packet %a" pp_error e);
-                      Ok (state, out, None)
-                  | Ok (session, transport) ->
-                      let state = { state with session }
-                      and ch = { ch with transport } in
-                      Ok (set_ch state ch, out, act))
-              | ( _control_op,
+                  let acts =
+                    Option.fold payload ~none:acts ~some:(fun p ->
+                        `Payload p :: acts)
+                  in
+                  (state, out, acts)
+              | Packet.Hard_reset_client_v3, _ ->
+                  Error (`No_transition (ch, op, payload))
```

Comment:
this is slightly dubious, but preserves the earlier semantics

Reply:
esp. that there haven't been any verification of the packet taken place makes it dubious (opens an attacker to send us a hard_reset_client_v3 to result in an error state).

Reply:
Ah yes for UDP we should maybe just drop it

Reply:
Right. This is maybe okay for TCP, but for UDP we should at least try to authenticate before returning ``Error `No_transition _``.

Maybe we should do a TCP/UDP error semantics overhaul at some point...

Reply:
I think the UDP error semantics needs revision, but this is fine for now.